compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts

```diff
@@ -0,0 +1,512 @@
+      prunedScopes.add(scopeStartInfo.scope.id);
+    }
+    for (const instr of block.instructions) {
+      for (const place of eachInstructionOperand(instr)) {
+        handlePlace(place);
+      }
+      handleInstruction(instr);
+    }
+
+    for (const place of eachTerminalOperand(block.terminal)) {
+      handlePlace(place);
+    }
+  }
+  return usedOutsideDeclaringScope;
+}
+
+type Decl = {
+  id: InstructionId;
+  scope: Stack<ReactiveScope>;
+};
+
+class Context {
+  #temporariesUsedOutsideScope: Set<DeclarationId>;
+  #declarations: Map<DeclarationId, Decl> = new Map();
+  #reassignments: Map<Identifier, Decl> = new Map();
+  // Reactive dependencies used in the current reactive scope.
+  #dependencies: Stack<Array<ReactiveScopeDependency>> = empty();
+  /*
+   * We keep a sidemap for temporaries created by PropertyLoads, and do
+   * not store any control flow (i.e. #inConditionalWithinScope) here.
+   *  - a ReactiveScope (A) containing a PropertyLoad may differ from the
+   *    ReactiveScope (B) that uses the produced temporary.
+   *  - codegen will inline these PropertyLoads back into scope (B)
+   */
+  #properties: ReadonlyMap<Identifier, ReactiveScopeDependency>;
+  #temporaries: ReadonlyMap<Identifier, Identifier>;
+  #scopes: Stack<ReactiveScope> = empty();
+  deps: Map<ReactiveScope, Array<ReactiveScopeDependency>> = new Map();
+
+  constructor(
+    temporariesUsedOutsideScope: Set<DeclarationId>,
+    temporaries: ReadonlyMap<Identifier, Identifier>,
+    properties: ReadonlyMap<Identifier, ReactiveScopeDependency>,
+  ) {
+    this.#temporariesUsedOutsideScope = temporariesUsedOutsideScope;
+    this.#temporaries = temporaries;
+    this.#properties = properties;
+  }
+
+  enterScope(scope: ReactiveScope): void {
+    // Set context for new scope
+    this.#dependencies = this.#dependencies.push([]);
+    this.#scopes = this.#scopes.push(scope);
+  }
+
+  exitScope(scope: ReactiveScope, pruned: boolean): void {
+    // Save dependencies we collected from the exiting scope
+    const scopedDependencies = this.#dependencies.value;
+    CompilerError.invariant(scopedDependencies != null, {
+      reason: '[PropagateScopeDeps]: Unexpected scope mismatch',
+      loc: scope.loc,
+    });
+
+    // Restore context of previous scope
+    this.#scopes = this.#scopes.pop();
+    this.#dependencies = this.#dependencies.pop();
+
+    /*
+     * Collect dependencies we recorded for the exiting scope and propagate
+     * them upward using the same rules as normal dependency collection.
+     * Child scopes may have dependencies on values created within the outer
+     * scope, which necessarily cannot be dependencies of the outer scope.
+     */
+    for (const dep of scopedDependencies) {
+      if (this.#checkValidDependency(dep)) {
+        this.#dependencies.value?.push(dep);
+      }
+    }
+
+    if (!pruned) {
+      this.deps.set(scope, scopedDependencies);
+    }
```

Comment:
Note that here, we store a scope's dependencies to later merge with results from `collectHoistablePropertyLoads`. The ordering doesn't really matter here, as both scope dependencies and non-null objects are read-only.

If this is confusing, I can move the merging logic here